libraries/lib-numeric-formats/NumericConverterFormats.cpp

```diff
@@ -0,0 +1,125 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file NumericConverterFormats.cpp
+
+ Dmitry Vedenko
+
+ **********************************************************************/
+
+#include "NumericConverterFormats.h"
+#include "NumericConverterRegistry.h"
+
+#include <cassert>
+#include <unordered_map>
+
+namespace
+{
+std::unordered_map<NumericConverterType, NumericFormatSymbol> DefaultSymbols;
```

Comment:
The map is guaranteed to be initialized before the first use and only once. There is precisely zero need for lazy initialization here.

Reply:
Then once again, you went to the trouble to make a registry, without taking actual advantage of a registry, so what was the point?

Reply:
Suppose we wanted the well-known frequency and bandwidth formats to be kept separate from time formats, which might make some sense for better separation of concernts.  Then there would be some scattering of the cases of this new small registry, and then what I said about the static variable would be relevant.

This need for a static defined within a function is such a frequency occurring pattern with registrations that I had to mention it.

Reply:
Consider this (please, not "blocking!") suggestion about what would seem to me a better organization.

- `FrequencyConverterFormats_`, and what depends on it, `parsedFrequency`, move out of `ParsedNumericConverterFormatter.cpp` into some new file.
- Then `NumericConverterType_FREQUENCY`, used by `parsedFrequency`, also moves its definition into that new file and its declaration into the corresponding header.  That requires `frequencyDefault` to move too.
- Now `NumericConverterType.*` don't mention frequency cases at all.
- Do similar for bandwidths: collect all that is special to bandwidth formats in one file.
- Even the time-specific formats and registrators might be separated from `ParsedNumericFormatConverter`.  Then that file is responsible for defining and interpreting the little language, but is (almost) agnostic about special case details.  (But there are still some legacy oddities about sample rates and NTSC frames.)
- Then `NumericConverterType.*` aren't needed.
- Dependency inversion and open endedness are really achieved:  the special cases (time, frequency, bandwidth) depend on the general case, each special case concentrated into one source file.  The open-endedness of the registration system is really used to advantage to get this.

All that would seem the more natural organization of things to me, and when you first mentioned you were using registrations, I hoped that this was where it was tending.

Reply:
> So I think beats and triplets should move out of SnapUtils.cpp and join beatsTime in one file.

Absolutely not.
 
> are really achieved

There is absolutely no difference to open-endedness in where those initializations reside. They are in the same file for the sake of reducing the number of changes so your own rule is followed.

Reply:
> Then NumericConverterType_FREQUENCY, used by parsedFrequency, also moves its definition into that new file

This will make the `NumericConverter` to be an absolute nightmare to use. It does expect a fixed type by design, there is no audacity-user choice in that. Making it an Identifier made things bad. Your IDE no longer can help you. Your compiler no longer can help you.  Scattering the possible values across the files will leave the developer without any clue about how to use `NumericConverter` or `NumericTextCtrl`. Enum are not always bad. Not everything need to be extensible. 

Speaking of software design - `NumericTextCtrl` should not be a subclass of `NumericConverter`. They are unrelated types. Considering that you chose not to make methods like ControlsToValues to be virtual - they cannot be even interchanged safely.

Reply:
> Speaking of software design - `NumericTextCtrl` should not be a subclass of `NumericConverter`. They are unrelated types. Considering that you chose not to make methods like ControlsToValues to be virtual - they cannot be even interchanged safely.

No disagreement on that -- and that it still is doesn't reflect any particular wish of mine.

I think you will discover that James separated the classes first, very long ago.